server/app/controllers/applicant/ApplicantProgramBlocksController.java

```diff
@@ -513,18 +525,37 @@ private CompletionStage<Result> renderErrorOrRedirectToNextBlock(
               .createAddressQuestion();
 
       if (addressQuestion.needsAddressCorrection()) {
-        return applicantService
-            .getAddressSuggestionGroup(thisBlockUpdated)
-            .thenApplyAsync(
-                addressSuggestionGroup -> {
-                  ImmutableList<AddressSuggestion> suggestions =
-                      addressSuggestionGroup.getAddressSuggestions();
-                  String json = addressSuggestionJsonSerializer.serialize(suggestions);
-
-                  Boolean isEligibilityEnabledOnThisBlock =
-                      thisBlockUpdated.getLeafAddressNodeServiceAreaIds().isPresent();
-
-                  return ok(addressCorrectionBlockView.render(
+        AddressSuggestionGroup addressSuggestionGroup =
```

Comment:
I'm making the call to get the address suggestions synchronous here. I feel like that isn't ideal (??) but I couldn't figure out how else to have the user paths divide. 

The key here is that if there is a matching suggestion, we want to save it to the backend and then move on to the next screen without showing the address correction screen (calling `confirmAddressWithSuggestions` accomplishes this because it invokes `renderErrorOrRedirectToNextBlock` again, this time with a corrected address). If we don't find a matching suggestion, we want to show the user the address correction screen.

Advice welcome!

Reply:
To make this asynchronous, return the completion stage returned by `applicantService.getAddressSuggestionGroup` move the remainder of the method into a callback on it:

```java
return applicantService.getAddressSuggestionGroup(thisBlockUpdated)
  .thenComposeAsync((AddressSuggestionGroup addressSuggestionGroup) -> {
        ImmutableList<AddressSuggestion> suggestions =
            addressSuggestionGroup.getAddressSuggestions();
        // etc
      },
      httpExecutionContext.current());

```

Reply:
@bion that's how it is [currently](https://github.com/civiform/civiform/blob/main/server/app/controllers/applicant/ApplicantProgramBlocksController.java#L516) and the issue I ran into there is the branching code paths. returning `return ok(addressCorrectionBlockView.render(...` is all well and good, but if we don't want to render the correction screen, we need to call `confirmAddressWithSuggestions` which has a `CompletionStage<Result>` type which `thenComposeAsync` doesn't like. another thought i had was to refactor `confirmAddressWithSuggestions` so it returns a `Result` but I couldn't figure out how to do that with the code contained in that method.